### PR TITLE
rustdoc: Document lack of object safety on affected traits

### LINKED ISSUE #85138: Show whether a trait is object-safe on its document page
So that it is easy for user to recognize object-safe traits.

### PR DESCRIPTION
**Pull request by poliorcetics**
Closes #85138

I saw the issue didn't have any recent activity, if there is another MR for it I missed it.

I want the issue to move forward so here is my proposition.

It takes some space just before the "Implementors" section and only if the trait is **not** object
safe since it is the only case where special care must be taken in some cases and this has the
benefit of avoiding generation of HTML in (I hope) the common case.

### CODE REVIEW

**Comment by poliorcetics** [comment]
Screenshot of the new 'Object Safety' section just above the 'Implementors' one on the Default trait, also showing the link to the section in the sidebar

File: src/librustdoc/html/render/print_item.rs
Comment (fmease) [inline src/librustdoc/html/render/print_item.rs:971]:
```suggestion
            "<div class=\"object-safety-info\">This trait is <b>not</b> \
            <a href=\"https://doc.rust-lang.org/stable/reference/items/traits.html#object-safety\">\
            object safe</a>.</div>",
```

**Review by fmease — CHANGES_REQUESTED** [review]
Thanks! Please add `object-safety` to the list of IDs in file `src/librustdoc/html/markdown.rs`, function `init_id_map`.

File: src/librustdoc/html/render/print_item.rs
Comment (GuillaumeGomez) [inline src/librustdoc/html/render/print_item.rs:967]: Why using `concat` here?

File: src/librustdoc/html/render/print_item.rs
Comment (GuillaumeGomez) [inline src/librustdoc/html/render/print_item.rs:962]: Shouldn't the condition be the opposite? Don't we want to display this information in case the trait is object safe? Or did I miss something?

File: src/librustdoc/html/render/sidebar.rs
Comment (GuillaumeGomez) [inline src/librustdoc/html/render/sidebar.rs:213]: Same question as above.

File: src/librustdoc/html/render/print_item.rs
Comment (GuillaumeGomez) [reply src/librustdoc/html/render/print_item.rs:962]: Ok nevermind. Just realized it were displaying the opposite information.

**Review by notriddle — APPROVED** [review]

**Review by GuillaumeGomez — APPROVED** [review]

**Review by fmease — APPROVED** [review]

**Comment by GuillaumeGomez** [comment]
Since it's making UI changes, let's start an FCP.

@rfcbot fcp start

**Comment by GuillaumeGomez** [comment]
Wrong command...

@rfcbot fcp merge

File: src/librustdoc/html/render/print_item.rs
Comment (GuillaumeGomez) [inline src/librustdoc/html/render/print_item.rs:972]: For the anchor:

```suggestion
            &format!(
                "<h3 id=\"{id}\">Object safety<a href=\"#{id}\" class=\"anchor\">§</a></h3>\
                 This trait is <b>not</b> \
                 <a href=\"{base}/reference/items/traits.html#object-safety\">\
                 object safe</a>.",
                base = crate::clean::utils::DOC_RUST_LANG_ORG_CHANNEL,
                id = "object-safety",
             ),
```

Then in the test, just check for the ID presence. One thing you'll need to update too will be to add `object-safety` ID into `init_id_map` (in `html/markdown.rs`).

Comment (poliorcetics) [reply src/librustdoc/html/render/print_item.rs:972]: `write_small_section_header` already does all of this, I just added a test for the ID

**Comment by poliorcetics** [comment]
@Nemo157 I think this PR is only waiting for review/approval, if you have the time

**Comment by GuillaumeGomez** [comment]
Soory, I completely missed this FCP when I was filling the agenda. I added it in the rustdoc agenda to be discussed next meeting if it hasn't been approved until then.

**Comment by GuillaumeGomez** [comment]
Thanks for the improvement!

@bors r+ rollup

**Comment by obi1kenobi** [comment]
Would it be possible to expose this data in rustdoc JSON as well?

It would be very useful to `cargo-semver-checks`, since object safety is public API and upholding it is a tricky semver hazard.

**Comment by GuillaumeGomez** [comment]
Sure.

**Comment by obi1kenobi** [comment]
Thank you! 🚀

Another thing that would be super useful to have (it would unlock [many](https://github.com/obi1kenobi/cargo-semver-checks/issues/294) [new](https://github.com/obi1kenobi/cargo-semver-checks/issues/441) [lints](https://github.com/obi1kenobi/cargo-semver-checks/issues/5)) is knowing whether a trait is sealed or not. This is not currently shown in either HTML or JSON output. _In theory_ it is possible to compute given sufficient rustdoc JSONs (including the stdlib), but it's quite hard to do since we don't have all of the compiler's infrastructure available and there are so many different ways to seal a trait ([partial list here](https://predr.ag/blog/definitive-guide-to-sealed-traits-in-rust/)).

**Comment by GuillaumeGomez** [comment]
So a trait that has a requirement for a local trait (based on [this blog post](https://predr.ag/blog/definitive-guide-to-sealed-traits-in-rust/)). One question I have is: if the required private trait has a blanket impl, is the trait still sealed?

**Comment by obi1kenobi** [comment]
> One question I have is: if the required private trait has a blanket impl, is the trait still sealed?

The trait is still sealed if the blanket impl has a bound on a trait that is itself sealed. Otherwise, it is not sealed -- it can be impl'd outside its defining crate.

> So a trait that has a requirement for a local trait (based on [this blog post](https://predr.ag/blog/definitive-guide-to-sealed-traits-in-rust/)).

There are a few more ways to make a trait not implementable outside its crate. I believe this to be an exhaustive list:
- if the trait has a sealed supertrait (sealed in any way, not just pub-in-priv)
- if the trait has a "sealed method" without a default impl, in any of the ways that allow sealing a method:
  - a pub-in-priv type as a parameter or return type (described [here](https://predr.ag/blog/definitive-guide-to-sealed-traits-in-rust/))
  - a sealed trait as a bound (described [here](https://old.reddit.com/r/rust/comments/12cj6as/a_definitive_guide_to_sealed_traits_in_rust/jf21zsm/))
- if the trait has an associated type with a sealed trait as a bound
- if the trait has an associated const of a pub-in-priv type that doesn't have a default

This is why it's so hard to determine whether a trait is sealed or not without all the compiler machinery being present 😅

**Comment by GuillaumeGomez** [comment]
Sounds like it would be quite tricky to check. However I think the information would be important to have. Can you open a new issue so the rustdoc team can discuss about it please? (With all the explanations too :) )

**Comment by Nemo157** [comment]
https://rust-lang.github.io/rfcs/3323-restrictions.html adds native syntax for sealed traits, it might be better to only support that (which will presumably be carried into the rustdoc-json somehow) and just require users of the existing hacks to migrate (which should be non-breaking, other than MSRV).

**Comment by obi1kenobi** [comment]
Opened https://github.com/rust-lang/rust/issues/119280 with more precise descriptions of when a trait is sealed.

To be honest, even if that RFC were to become available on stable tomorrow, I don't expect users of the existing approaches will migrate anytime soon. The linked issue has details, but TL;DR maintainers have minimal incentive to change code that from their perspective 100% works fine just to adopt a new syntax.

**Pull request merged**
